src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -0,0 +1,774 @@
+
+        // It should be impossible for multiple threads to access these at the same time because of the semaphore we use.
+        private readonly SemaphoreSlim _rowsLock;
+        private IReadOnlyList<IReadOnlyDictionary<string, string>> _rows;
+        private int _leaseRenewalCount;
+        private State _state = State.CheckingForChanges;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SqlTableChangeMonitor{T}" />> class.
+        /// </summary>
+        /// <param name="connectionString">The SQL connection string used to connect to the user's database</param>
+        /// <param name="userTableId">The OBJECT_ID of the user table whose changes are being tracked on</param>
+        /// <param name="userTableName">The name of the user table</param>
+        /// <param name="userFunctionId">The unique ID that identifies user function</param>
+        /// <param name="workerTableName">The name of the worker table</param>
+        /// <param name="userTableColumns">List of all column names in the user table</param>
+        /// <param name="primaryKeyColumns">List of primary key column names in the user table</param>
+        /// <param name="executor">Used to execute the user's function when changes are detected on "table"</param>
+        /// <param name="logger">Ilogger used to log information and warnings</param>
+        public SqlTableChangeMonitor(
+            string connectionString,
+            int userTableId,
+            SqlObject userTable,
+            string userFunctionId,
+            string workerTableName,
+            IReadOnlyList<string> userTableColumns,
+            IReadOnlyList<string> primaryKeyColumns,
+            ITriggeredFunctionExecutor executor,
+            ILogger logger)
+        {
+            _ = !string.IsNullOrEmpty(connectionString) ? true : throw new ArgumentNullException(nameof(connectionString));
+            _ = !string.IsNullOrEmpty(userTable.FullName) ? true : throw new ArgumentNullException(nameof(userTable));
+            _ = !string.IsNullOrEmpty(userFunctionId) ? true : throw new ArgumentNullException(nameof(userFunctionId));
+            _ = !string.IsNullOrEmpty(workerTableName) ? true : throw new ArgumentNullException(nameof(workerTableName));
+            _ = userTableColumns ?? throw new ArgumentNullException(nameof(userTableColumns));
+            _ = primaryKeyColumns ?? throw new ArgumentNullException(nameof(primaryKeyColumns));
+            _ = executor ?? throw new ArgumentNullException(nameof(executor));
+            _ = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            this._connectionString = connectionString;
+            this._userTableId = userTableId;
+            this._userTable = userTable;
+            this._userFunctionId = userFunctionId;
+            this._workerTableName = workerTableName;
+            this._userTableColumns = primaryKeyColumns.Concat(userTableColumns.Except(primaryKeyColumns)).ToList();
```

Comment:
Is this extra logic actually necessary? If the user table columns already has the primary key columns (which I would expect they do) why do we need to remove them and then concat them back on?

Reply:
This is done to have all primary key columns before all non-primary-key columns. I am not sure, but it might be required in one of the SQL queries. I will check this out.

Reply:
Alrighty, if it ends up being necessary make sure to add a comment explaining that

Reply:
The ordering of columns is no longer needed, probably because of the changes we made to how query text is constructed. I will simplify the statement.